### PR TITLE
Restructuring functionalities to Import an API/API Product with new schemas with relevant DTOs for APICTL

### PR DESCRIPTION
**Pull request by wasuradananjith**
## Purpose
Fixes: https://github.com/wso2/product-apim-tooling/issues/324
Fixes: https://github.com/wso2/product-apim-tooling/issues/158

## Goals
Restructure the code segments related to importing an API/API Product by introducing new schemas via reading required DTOs.

## Approach
- Created Publisher REST API resources to import an API/API Product. The resources are as follows.
  1. POST apis/import
  2. POST api-products/import
- The old implementations of importing APIs and API Products have been modified to read DTOs from the importing files, instead of models.
- The new structure introduced in [1] has been further modified with the below changes.
   - The “Meta-information” folder was renamed as “Definitions”. 
   - Brought the api.yaml to the root level.
   - When exporting documents, currently the document name will be the name of the individual document directory. (Example, Doc1 in [1] is the document name of the documentation related to that directory). The document directory name has been further validated and replaced the unwanted characters before exporting it as the directory name for the safe side.
   ```
   └── PizzaShackAPI-1.0.0
       ├── api.yaml
       ├── Client-certificates
       │   ├── Alias1.crt
       │   ├── Alias2.crt
       │   └── client_certificates.yaml
       ├── Docs
       │   ├── Doc1
       │   │   ├── mydoc.pdf
       │   │   └── document.yaml
       │   ├── Doc2
       │   │   ├── Doc2
       │   │   └── document.yaml
       │   ├── Doc3
       │   │   ├── Doc3
       │   │   └── document.yaml
       │   └── Doc4
       │       └── document.yaml
       ├── Endpoint-certificates
       │   ├── Alias3.crt
       │   ├── Alias4.crt
       │   └── endpoint_certificates.yaml
       ├── Image
       │   └── icon.png
       ├── Definitions
       │   └── swagger.yaml
       └── Sequences
           ├── fault-sequence
           │   ├── sequence.yaml
           │   └── debug_json_fault.xml
           ├── in-sequence
           │   ├── sequence.yaml
           │   └── Custom
           │       └── sample-n-sequence.xml
           └── out-sequence
                      ├── sequence.yaml
                      └── log_out_message.xml
   ```
- A known issue of requiring the thumbnail name to be “icon” when importing a thumbnail image along with the API project has been fixed
- Fixed exception handling during the import according to [2].

## User stories
The users can use the newly introduced Publisher REST API resources to import an API/API Product with the new schema structure.

## Documentation
Documentation changes need to be done according to the structure changes.

## Related PRs
https://github.com/wso2/carbon-apimgt/pull/9406

## Test environment
- JDK 1.8.0_251
- Ubuntu 20.04.1 LTS

[1] https://github.com/wso2/carbon-apimgt/pull/9406
[2] https://github.com/wso2/carbon-apimgt/wiki/Exception-handling-in-REST-APIs

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java [hunk 1]
```diff
@@ -31,36 +34,62 @@
...
+     * @param fileInputStream     Input stream from the REST request
+     *                            (This will not be null when importing dependent APIs with API Products)
+     * @param preserveProvider    User choice to keep or replace the API Product provider
+     * @param importAPIs          Whether to import the dependent APIs or not.
+     * @param overwriteAPIProduct Whether to update the API Product or not. This is used when updating already existing API Products.
+     * @param overwriteAPIs       Whether to update the dependent APIs or not. This is used when updating already existing dependent APIs of an API Product.
+     * @@return Imported API Product
+     */
+    public APIProduct importAPIProduct(InputStream fileInputStream, Boolean preserveProvider,
+            Boolean overwriteAPIProduct, Boolean overwriteAPIs, Boolean importAPIs, String[] tokenScopes)
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java:93]: Do we need to pass token scopes to interface

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java (hunk 1 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java:93]: The token scopes have been used in the updateApi function in PublisherCommonUtils.java. Earlier, the code segment related to this function was in the ApisApiServiceImpl.java. The token scopes are required when deciding whether a particular user can update a particular field of an API.

**Review by tharindu1st — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java [hunk 2]
```diff
@@ -31,36 +34,62 @@
...
+            boolean preserveStatus, boolean preserveDocs) throws APIManagementException, APIImportExportException;
+
+    /**
+     * Used to import an API artifact.
+     *
+     * @param fileInputStream  Input stream from the REST request
+     *                         (This will not be null when importing dependent APIs with API Products)
+     * @param preserveProvider Decision to keep or replace the provider
+     * @param overwrite        Whether to update the API or not
+     * @throws APIImportExportException If there is an error in importing an API
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java:75]: Wrong exception is mentioned here

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java [hunk 3]
```diff
@@ -31,36 +34,62 @@
...
+     * Used to import an API Product artifact.
+     *
+     * @param fileInputStream     Input stream from the REST request
+     *                            (This will not be null when importing dependent APIs with API Products)
+     * @param preserveProvider    User choice to keep or replace the API Product provider
+     * @param importAPIs          Whether to import the dependent APIs or not.
+     * @param overwriteAPIProduct Whether to update the API Product or not. This is used when updating already existing API Products.
+     * @param overwriteAPIs       Whether to update the dependent APIs or not. This is used when updating already existing dependent APIs of an API Product.
+     * @@return Imported API Product
+     */
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java:91]: Missing values in the java docs. Please add.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApiProductsApiServiceImpl.java [hunk 4]
```diff
@@ -726,77 +676,65 @@ public Response apiProductsGet(Integer limit, Integer offset, String query, Stri
...
+     * Import an API Product by uploading an archive file. All relevant API Product data will be included upon the creation of
+     * the API Product. Depending on the choice of the user, provider of the imported API Product will be preserved or modified.
+     *
+     * @param fileInputStream       UploadedInputStream input stream from the REST request
+     * @param fileDetail            File details as Attachment
+     * @param preserveProvider      User choice to keep or replace the API Product provider
+     * @param importAPIs            Whether to import the dependent APIs or not.
+     * @param overwriteAPIProduct   Whether to update the API Product or not. This is used when updating already existing API Products.
+     * @param overwriteAPIs         Whether to update the dependent APIs or not. This is used when updating already existing dependent APIs of an API Product.
+     * @return API Product import response
```
Comment (npamudika) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApiProductsApiServiceImpl.java:689]: Missing @param and @throws

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java (hunk 2 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java:75]: Fixed via https://github.com/wso2/carbon-apimgt/commit/eb9b7181c0f6cb5b9624602d91d28a7479bf551b

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java (hunk 3 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/importexport/ImportExportAPI.java:91]: Fixed via https://github.com/wso2/carbon-apimgt/commit/eb9b7181c0f6cb5b9624602d91d28a7479bf551b

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApiProductsApiServiceImpl.java (hunk 4 above)
Comment (wasuradananjith) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/impl/ApiProductsApiServiceImpl.java:689]: Fixed via https://github.com/wso2/carbon-apimgt/commit/eb9b7181c0f6cb5b9624602d91d28a7479bf551b

**Pull request merged**